File: app/src/main/java/com/lucagiorgetti/surprix/ui/doublelist/DoubleListViewModel.java
package com.lucagiorgetti.surprix.ui.doublelist;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback;
import com.lucagiorgetti.surprix.model.Surprise;
import com.lucagiorgetti.surprix.ui.BaseViewModel;
import com.lucagiorgetti.surprix.utility.DatabaseUtility;

import java.util.List;

public class DoubleListViewModel extends BaseViewModel {

    private MutableLiveData<List<Surprise>> allDoubleSurprises;

    public DoubleListViewModel(@NonNull Application application) {
        super(application);
        this.setLoading(false);
    }

    public MutableLiveData<List<Surprise>> getDoubleSurprises() {
        if (allDoubleSurprises == null) {
            allDoubleSurprises = new MutableLiveData<>();
            loadDoubleSurprises();
        }

        return allDoubleSurprises;
    }


    public void loadDoubleSurprises() {
        DatabaseUtility.getDoublesForUsername(new FirebaseListCallback<Surprise>() {
            @Override
            public void onStart() {
                setLoading(true);
            }

            @Override
            public void onSuccess(List<Surprise> doubleSurprises) {
                allDoubleSurprises.setValue(doubleSurprises);
                setLoading(false);
            }

            @Override
            public void onFailure() {
                allDoubleSurprises.setValue(null);
                setLoading(false);
            }
        });
    }

    public void addDouble(Surprise surprise, int position) {
        List<Surprise> list = allDoubleSurprises.getValue();
        list.add(position, surprise);
        allDoubleSurprises.setValue(list);
    }
}

File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/surpriseList/doublelist/DoubleListViewModel.kt
package com.lucagiorgetti.surprix.ui.mainfragments.surpriseList.doublelist

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.lucagiorgetti.surprix.SurprixApplication
import com.lucagiorgetti.surprix.model.Surprise
import com.lucagiorgetti.surprix.ui.BaseViewModel
import com.lucagiorgetti.surprix.ui.mainfragments.surpriseList.SurpiseListFirebaseCallback
import com.lucagiorgetti.surprix.utility.dao.DoubleListDao

class DoubleListViewModel(application: Application) : BaseViewModel(application) {
    private var allDoubleSurprises: MutableLiveData<MutableList<Surprise>> ? = null
    val doublesSurprisesCount = MutableLiveData<Int?>()

    init {
        setLoading(false)
    }

    fun getDoubleSurprises(): MutableLiveData<MutableList<Surprise>>{
        if (allDoubleSurprises == null) {
            allDoubleSurprises = MutableLiveData()
            loadDoubleSurprises()
        }
        return allDoubleSurprises!!
    }

    fun loadDoubleSurprises() {
        val surprises: MutableList<Surprise> = ArrayList()

        DoubleListDao(SurprixApplication.instance.currentUser?.username).getDoubles(object : SurpiseListFirebaseCallback {
            override fun onStart() {
                setLoading(true)
                doublesSurprisesCount.value = null
            }

            override fun onSuccess(doubleSurprises: Surprise) {
                surprises.add(doubleSurprises)
                allDoubleSurprises!!.value = surprises
                setLoading(false)
            }

            override fun onFailure() {
                allDoubleSurprises!!.value = ArrayList()
                setLoading(false)
            }

            override fun onNewData() {
                surprises.clear()
                allDoubleSurprises!!.value = surprises
            }

            override fun onCountDiscovered(count: Int) {
                doublesSurprisesCount.value = count
            }
        })
    }

    fun addDouble(surprise: Surprise, position: Int) {
        val list = allDoubleSurprises!!.value
        list!!.add(position, surprise)
        allDoubleSurprises!!.value = list
    }
}